### ISSUE TITLE
Questions about the inconsistent (slightly different) results obtained from multiple runs of nn.NLLLoss

### ISSUE THREAD

**Issue by MaxChanger** [pytorch/pytorch#68698]
## may be a Bug?  I'm not sure

My original purpose was to fix the randomness of a network, but I fixed all seeds, benchmark, deterministic, etc., but it still didn’t work. So I started to output the value of each structure step by step, and finally found that there is a certain randomness in the loss part. Even if I use pdb to debug and repeat the calculation many times, the results are slightly different.
Due to the existence of this part of randomness, after multiple iterations of backward, run train.py twice, the network prediction results may have a large deviation.

I'm not sure if this has anything to do with the precision that float can express, I hope to get some more reasonable explanations to clarify my confusion.

In order to easily reproduce this problem, I saved the prediction and label used in a certain iteration, [coda and numpy data is here](https://drive.google.com/file/d/1xHjAdK1YYJU6V-NMe7tVF_lsiSSzTUJB/view?usp=sharing), and used a simple code to reproduce.

```python
import numpy as np
import torch 
import torch.nn as nn

tmp_log = np.load("./tmp_log.npy")
labels = np.load("./proj_labels.npy")

log = torch.from_numpy(tmp_log).cuda()
labels = torch.from_numpy(labels).long().cuda()
loss = nn.NLLLoss().cuda()

for i in range(50):
    output = loss(log, labels)
    print(i, output.item())

```
Some details and output are as follows:

![image](https://user-images.githubusercontent.com/35401968/142752704-e4b4fd01-5ca4-4840-8a23-310853de309c.png)

## To Reproduce

1. Download the two tensor npy files I saved from this drive.google link. [coda and numpy data is here](https://drive.google.com/file/d/1xHjAdK1YYJU6V-NMe7tVF_lsiSSzTUJB/view?usp=sharing)
2. `python check_NLLLoss.py`
3. And then you will get an output similar to the image above.

## Expected behavior

I hope to get a fixed output to ensure the reproducibility of my network.

## Environment
I use PyTorch 1.7.0 or 1.5.0, at 3090 or 2080Ti, if necessary, I will provide more detailed information.

 - PyTorch Version : 1.7.0 or 1.5.0
 - OS (e.g., Linux): ubuntu 18.04
 - How you installed PyTorch (`conda`, `pip`, source):  conda
 - Python version: 3.7.7

cc @albanD @mruberry @jbschlosser @walterddr @ngimel

**Comment by albanD** [comment]
Hi,

Yes this is expected behavior.
This op is not deterministic by default see https://pytorch.org/docs/stable/generated/torch.use_deterministic_algorithms.html?highlight=deterministic#torch.use_deterministic_algorithms for more details on which ops are not deterministic and how to force them to be (even though it will be slower).

**Comment by MaxChanger** [comment]
@albanD Thanks a lot.
There is still a small question. 
Setting `torch.backends.cudnn.deterministic = True` and `torch.use_deterministic_algorithms(True) / torch.set_deterministic(True)` will make some normally-nondeterministic operations in the link's list act deterministically, and the difference is whether it will report an error or not.

Is this understanding correct or not?

**Comment by albanD** [comment]
Yes that's correct.
